tests/test_weighted_random_sampler.py

```diff
@@ -1,59 +1,146 @@
 import numpy as np
 import pytest
-from scipy.stats import ks_2samp
+from numpy.typing import NDArray
 
 from timemachine.lib import custom_ops
 
 pytestmark = [pytest.mark.memcheck]
 
 
-@pytest.mark.parametrize("seed", [2022])
-@pytest.mark.parametrize("size, num_samples", [(500, 1500), (1000, 3000)])
+def normalize_probabilities(vals: NDArray) -> NDArray:
+    return vals / np.sum(vals)
+
+
+@pytest.mark.parametrize("seed", [2024])
 @pytest.mark.parametrize("precision", [np.float32, np.float64])
-def test_random_sampler(seed, size, num_samples, precision):
+def test_segmented_random_sampler_validation(seed, precision):
+    klass = custom_ops.SegmentedWeightedRandomSampler_f32
+    if precision == np.float64:
+        klass = custom_ops.SegmentedWeightedRandomSampler_f64
+
+    max_vals_per_segment = 5
+    segments = 2
+
+    sampler = klass(max_vals_per_segment, segments, seed)
+
+    with pytest.raises(RuntimeError, match="number of segments don't match"):
+        sampler.sample([[1.0]])
+
+    with pytest.raises(RuntimeError, match="total values is greater than buffer size"):
+        sampler.sample([[1.0] * (max_vals_per_segment + 1) for _ in range(segments)])
+
+    with pytest.raises(RuntimeError, match="empty probability distribution not allowed"):
+        sampler.sample([[5.0, 3.0, 10.0], []])
+
+    with pytest.raises(RuntimeError, match="all values in log space must be finite and non-negative"):
+        sampler.sample([[np.inf], [np.inf]])
+
+    with pytest.raises(RuntimeError, match="all values in log space must be finite and non-negative"):
+        sampler.sample([[-np.inf], [-np.inf]])
+
+    with pytest.raises(RuntimeError, match="all values in log space must be finite and non-negative"):
+        sampler.sample([[np.nan], [np.nan]])
+
+    with pytest.raises(RuntimeError, match="all values in log space must be finite and non-negative"):
+        sampler.sample([[1.0], [-0.1]])
+
+    with pytest.raises(RuntimeError, match="all values in log space must be finite and non-negative"):
+        sampler.sample([[1.0], [0.0]])
+
+
+@pytest.mark.parametrize("seed", [2024, 2025, 2026, 2027])
+@pytest.mark.parametrize("size, num_samples", [(50, 6000), (100, 20000)])
+@pytest.mark.parametrize("precision", [np.float32, np.float64])
+def test_segmented_random_sampler(seed, size, num_samples, precision):
     rng = np.random.default_rng(seed)
 
     # Get a random probability vector.
-    probs = rng.exponential(scale=1 / 250, size=size)
-    prob_total = np.sum(probs)
-    probs /= prob_total
-    # Sort the probabilities to easily visualize
-    probs = np.sort(probs)[::-1]
+    alpha = rng.normal(loc=10.0, size=size)
+    probs = rng.dirichlet(alpha, size=num_samples)
```

Comment:
Based on a quick check, it looks like this will generate `num_samples` instances that are each very close to uniform.

(Sorry, I said this part of my suggestion backwards "make weights in each of the random instances much further from uniform (i.e. sample each row of probs from dirichlet, with high concentration parameter)")

To make the test more challenging, might use something like the following, where each row will be more or less "spiky." (for rows where alpha is high, the row will be closer to uniform, and for rows where alpha is low, the row will be closer to "one-hot")

```
probs = []
for _ in range(num_samples):
    # some rows will be "spiky" (small alpha), some rows will be more uniform (large alpha)
    alpha = rng.uniform(0.01, 1.0) * np.ones(size) 
    prob_vec = rng.dirichlet(alpha)

    # output will be normalized: maybe mess that up for testing
    # unnormalized_prob_vec = prob_vec * rng.uniform(0.1, 10.0)
    probs.append(prob_vec)
probs = np.array(probs)
```

I may be misunderstanding though -- should the test exercise the case where each of the segments describes a very different distribution from all the other segments?

Reply:
> should the test exercise the case where each of the segments describes a very different distribution from all the other segments

I don't think so? The distribution of log weights tends to be relatively uniform. But reading this again I should modify the test to sample the same distribution repeatedly (call sample multiple times) and compare the percentages. Else if you get spiky distributions you either miss values entirely or have large mismatches.

Reply:
Modified the logic you had suggested (to avoid generating probabilities of 0.0) and now sample repeatedly. https://github.com/proteneer/timemachine/pull/1235/commits/3a8e4ddb071e294e1f046f11a815b413f93053e5